cpp/src/arrow/util/memory-pool.cc

```diff
@@ -74,15 +74,22 @@ Status InternalMemoryPool::Allocate(int64_t size, uint8_t** out) {
   return Status::OK();
 }
 
-int64_t InternalMemoryPool::bytes_allocated() const {
+uint64_t InternalMemoryPool::bytes_allocated() const {
   std::lock_guard<std::mutex> guard(pool_lock_);
   return bytes_allocated_;
 }
 
-void InternalMemoryPool::Free(uint8_t* buffer, int64_t size) {
+Status InternalMemoryPool::Free(uint8_t* buffer, int64_t size) {
   std::lock_guard<std::mutex> guard(pool_lock_);
+  if (bytes_allocated_ < size) {
+    std::stringstream ss;
+    ss << "free of size " << size << " larger than allocated bytes " << bytes_allocated_ << " failed ";
+    return Status::Invalid(ss.str());
+  }
```

Comment:
I agree with DCHECK.  I think we should be rethinking the API a little bit here in general (outside the scope of this code review).  It would be nice to not have to explicitly carry around the size to deallocate, also having a explicit realloc method would be useful, since array growing is tightly coupled the allocator used (there is an open JIRA about this).   Doing the malloc trick of storing size in it in the memory locations prior to the returned pointer could make sense, but given the alignment requirements might waste too much space.  Another alternative is keeping a dictionary of memory location to size.